{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to\n",
      "[nltk_data]     /home/karlkeat/nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import logging\n",
    "import sys\n",
    "import os\n",
    "import openai\n",
    "from IPython.display import Markdown, display\n",
    "import nltk\n",
    "nltk.download(\"stopwords\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "metadata": {},
   "outputs": [],
   "source": [
    "from llama_index.llms.openai import OpenAI\n",
    "from llama_index.embeddings.openai import OpenAIEmbedding\n",
    "from llama_index.core import Settings\n",
    "\n",
    "# Set OpenAI API Key\n",
    "openai.api_key = os.environ.get(\"OPENAI_API_KEY\")\n",
    "\n",
    "# Add helicone API key\n",
    "openai.api_base = \"http://oai.hconeai.com/v1\"\n",
    "headers = {\n",
    "            \"Helicone-Auth\": f\"Bearer {os.environ.get('HELICONE_API_KEY')}\",\n",
    "            \"Helicone-Cache-Enabled\": \"true\"\n",
    "          }\n",
    "\n",
    "llm = OpenAI(temperature=0.1, model=\"gpt-3.5-turbo\", api_base=\"http://oai.hconeai.com/v1\", api_key=os.environ[\"OPENAI_API_KEY\"], default_headers=headers)\n",
    "embed_model = OpenAIEmbedding(temperature=0.1, model=\"text-embedding-3-small\", api_base=\"http://oai.hconeai.com/v1\", api_key=os.environ[\"OPENAI_API_KEY\"], default_headers=headers)\n",
    "\n",
    "Settings.llm = llm\n",
    "Settings.embed_model = embed_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build Clinical Guidelines Retriever"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "metadata": {},
   "outputs": [],
   "source": [
    "logging.basicConfig(stream=sys.stdout, level=logging.INFO)\n",
    "logging.getLogger().addHandler(logging.StreamHandler(stream=sys.stdout))\n",
    "\n",
    "from llama_index.core import (\n",
    "    VectorStoreIndex,\n",
    "    SimpleDirectoryReader,\n",
    "    load_index_from_storage,\n",
    "    StorageContext,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n"
     ]
    }
   ],
   "source": [
    "# Run during first time setup only\n",
    "cpic_guideline_documents = SimpleDirectoryReader(\"./cpic_guidelines\").load_data()\n",
    "cpic_guideline_index = VectorStoreIndex.from_documents(cpic_guideline_documents)\n",
    "cpic_guideline_index.set_index_id(\"cpic_guidelines_vector_index\")\n",
    "cpic_guideline_index.storage_context.persist(\"./vector_index_storage\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:llama_index.core.indices.loading:Loading indices with ids: ['cpic_guidelines_vector_index']\n",
      "Loading indices with ids: ['cpic_guidelines_vector_index']\n",
      "Loading indices with ids: ['cpic_guidelines_vector_index']\n",
      "Loading indices with ids: ['cpic_guidelines_vector_index']\n",
      "Loading indices with ids: ['cpic_guidelines_vector_index']\n",
      "Loading indices with ids: ['cpic_guidelines_vector_index']\n"
     ]
    }
   ],
   "source": [
    "# rebuild storage context\n",
    "storage_context = StorageContext.from_defaults(persist_dir=\"vector_index_storage\")\n",
    "# load index\n",
    "cpic_guideline_index = load_index_from_storage(storage_context, index_id=\"cpic_guidelines_vector_index\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[NodeWithScore(node=TextNode(id_='ada3a8d8-6a4f-48aa-9f31-e0b2e8327c8d', embedding=None, metadata={'page_label': '318', 'file_name': '23698643.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/23698643.pdf', 'file_type': 'application/pdf', 'file_size': 250786, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, excluded_embed_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], excluded_llm_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='4eecab71-b9ff-4ae7-8aab-cc19d8cbb949', node_type=<ObjectType.DOCUMENT: '4'>, metadata={'page_label': '318', 'file_name': '23698643.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/23698643.pdf', 'file_type': 'application/pdf', 'file_size': 250786, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='24586dccaa4b31e66cceb6154b9e982cb08b3973b6529153aaedc8b27e8ed4c1'), <NodeRelationship.PREVIOUS: '2'>: RelatedNodeInfo(node_id='17291f46-877c-4157-b763-56b5c0f7bd2a', node_type=<ObjectType.TEXT: '1'>, metadata={'page_label': '317', 'file_name': '23698643.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/23698643.pdf', 'file_type': 'application/pdf', 'file_size': 250786, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='5d1624d2c42a8f1c390d194a0d201250a9d167ccd331049b609e18de2c49b95a'), <NodeRelationship.NEXT: '3'>: RelatedNodeInfo(node_id='7d5d1af6-09f5-48db-81c0-6304e4952a0e', node_type=<ObjectType.TEXT: '1'>, metadata={}, hash='9f23e0cf790091a6965056873a84e3f349a1368645facb7d8cd8cc9bc9c6bbfa')}, text='CPIC UPdate\\nOnly 15% of the prodrug is available for transformation to the \\nactive agent; the remaining 85% is hydrolyzed by esterases to \\ninactive forms. Conversion of clopidogrel to its active metabo-\\nlite requires two sequential oxidative steps involving several \\nCYP enzymes (e.g., CYP1A2, CYP2B6, CYP2C9, CYP2C19, \\nand CYP3A4/5; Supplementary Figure S1 online). One small \\nassociation study suggested that paraoxonase 1 (PON1) may \\nalso be involved in clopidogrel activation;3 however, this finding \\nand the reported association between PON1 variant alleles and \\nclopidogrel response variability was not replicated in multiple \\nrecent publications (for review, see ref. 4).\\nGENE: CYP2C19\\nBackground\\nA gene summary on CYP2C19 has recently been published5 and \\nis available online at PharmGKB: http://www.pharmgkb.org/\\ngene/PA124#tabview=tab3&subtab=31. The hepatic CYP2C19 \\nenzyme contributes to the metabolism of a large number of clin-\\nically relevant drugs such as antidepressants, benzodiazepines, \\nmephenytoin, some proton pump inhibitors, and clopidogrel. \\nLike many other CYP450  superfamily members, the CYP2C19  \\ngene is highly polymorphic, with >25 known variant alleles (http://www.cypalleles.ki.se/cyp2c19.htm) (Supplementary \\nTables S1 and  S2 online).5 The wild-type CYP2C19*1 allele is \\nassociated with functional CYP2C19-mediated metabolism. The \\nmost common CYP2C19 loss-of-function allele is *2 (c.681G>A; \\nrs4244285), with allele frequencies of ~15% in Caucasians and \\nAfricans, and 29–35% in Asians (Supplementary Tables S3   \\nand S4 online). Other CYP2C19 variant alleles with reduced or \\nabsent enzymatic activity have been identified (e.g., *3 –*8); how -\\never, their frequencies are typically below 1%, with the exception \\nof CYP2C19*3 (c.636G>A; rs4986893) in Asians (with an allele \\nfrequency of 2–9%; Supplementary Tables S3  and  S4 online).\\nWith respect to clopidogrel pharmacodynamics, CYP2C19  \\nloss-of-function alleles are inherited as autosomal codominant \\ntraits because heterozygotes (e.g., *1/*2 and *1/*3) have platelet \\nresponsiveness to clopidogrel that lies between homozygous wild-type (i.e., *1/*1) and loss-of-function allele homozy-\\ngotes or compound heterozygotes (e.g.,. *2/*2 and *2/*3).6–11 \\nTherefore, on the basis of identified CYP2C19 genotypes, indi-\\nviduals typically are categorized as extensive metabolizers (EMs; \\n*1/*1), intermediate metabolizers (IMs; e.g., *1/*2 and *1/ *3), or poor metabolizers (PMs; e.g., *2/*2 and *2/*3) Table 1 . The \\nfrequencies of CYP2C19 PMs are ~2–5% among Caucasians and \\nAfricans and ~15% in Asians (Supplementary Materials  and \\nMethods  online).\\nBy contrast, the common CYP2C19*17 allele (c.-806C>T; \\nrs12248560) results in increased activity as a consequence of enhanced transcription, with average multiethnic allele fre -\\nquencies of ~3–21% (Supplementary Tables  S3 and S4 online). \\nTherefore, individuals who carry this allele may be categorized as ultrarapid metabolizers (e.g., *17/*17) Table 1 . Some studies \\nindicate that this allele results in enhanced platelet inhibition \\nand clopidogrel response10,12–14 and perhaps an increased risk \\nof bleeding complications.15,16 However, other studies have not \\nidentified an effect of CYP2C19*17.11,17–20 Discordant results \\nin the literature may be due in part to the linkage disequilib-\\nrium that exists between *17 and *2.', start_char_idx=0, end_char_idx=3372, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=0.7751280209567996),\n",
       " NodeWithScore(node=TextNode(id_='296c6ac4-7823-4493-9f4a-fdb918365d2a', embedding=None, metadata={'page_label': '2', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, excluded_embed_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], excluded_llm_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='896f25dd-93de-4de0-a92e-4038eb7e61e4', node_type=<ObjectType.DOCUMENT: '4'>, metadata={'page_label': '2', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='695aaf994c5da0e06d1aed61b478c13b5a4258f3e5a2b72cf7e30e471c87990b'), <NodeRelationship.PREVIOUS: '2'>: RelatedNodeInfo(node_id='0625cf87-e45e-442b-a300-8ac47159be9f', node_type=<ObjectType.TEXT: '1'>, metadata={'page_label': '1', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='a048611c79376f118c29ae77d894d855a45c1e5c9fd6afb95f9859bf119cedd1'), <NodeRelationship.NEXT: '3'>: RelatedNodeInfo(node_id='214f0dc0-a99a-40cb-b042-927ba7833c6f', node_type=<ObjectType.TEXT: '1'>, metadata={}, hash='7470ae69ee2e1a703ce92c21696db6365128bdadee58862013d09c7ce9a1c20a')}, text='2 www.nature.com/cpttranslation\\n(c.636G>A; rs4986893) in Asians (2–9%; Supplementary Data  \\nonline).\\nCYP2C19*2 is inherited as an autosomal co-dominant trait; \\nplatelet responsiveness to clopidogrel in heterozygotes (*1/ *2) \\nlies somewhere between the responsiveness in individuals with \\nthe *1/*1 genotype and that in those with the *2/*2 genotype.2–7 \\nTherefore, based on identified CYP2C19  genotypes, individu-\\nals can be categorized as extensive metabolizers (e.g., *1/*1), \\nintermediate metabolizers (e.g., *1/*2), or poor metabolizers \\n(e.g., *2/*2). The frequencies of CYP2C19 poor metabolizers are \\n~2–5% among Caucasians and Africans and ~15% in Asians \\n(Supplementary Data online).\\nIn contrast, the CYP2C19*17 allele (c.-806C>T; rs12248560) \\nresults in increased activity as a consequence of enhanced \\ntranscription, with average multi-ethnic allele frequen-cies of ~3–21% (Supplementary Tables S3  and S4 online). \\nIndividuals who carry this allele may be categorized as ultra -\\nrapid metabolizers (e.g., *17/*17). Some studies indicate that \\nthis allele results in enhanced platelet inhibition and clopidog -\\nrel response,6,8–10 and possibly an increased risk of bleeding \\ncomplications.11 However, other studies have not identified an \\neffect of CYP2C19*17 (refs.  7,12,13), and adequate evidence \\nfor an independent effect of this allele on clinical outcomes \\nis lacking.\\nGenetic test interpretation.  Clinical genotyping tests are available \\nthat interrogate variant CYP2C19 alleles and predict an indi-\\nvidual’s CYP2C19 metabolizer phenotype (Supplementary \\nData online). Each named star (*) allele is defined by the geno-\\ntype at one or more specific single-nucleotide polymorphisms (Supplementary Table S1 online) and is associated with a level of enzyme activity (Supplementary Table S2 online). \\nTable\\n \\n1\\n \\nand Supplementary Table S5 online summarize the assign-\\nment of the likely CYP2C19 phenotype based on common star \\nallele diplotypes, and these assignments are used to link geno-types with personalized antiplatelet therapy.\\nAvailable genetic test options.  A number of academic and commer -\\ncial clinical laboratories offer genetic testing for CYP2C19*2 and \\nother CYP2C19 variant alleles (Supplementary Data online). \\nHowever, at the time of this writing, only the CYP2C19*2 allele \\nhas been adequately studied with respect to clinical outcomes \\non clopidogrel; other variants are too rare, have not been stud-ied, or have resulted in inconclusive findings.Incidental findings.  There are no diseases or conditions that have \\nbeen convincingly linked to variations in the CYP2C19 gene \\nindependent of drug metabolism and response.\\nOther considerations.  CYP2C19 loss-of-function alleles do not \\naccount for all of the variability in clopidogrel response. Some \\nstudies have implicated variants in other genes associated with \\nclopidogrel response (Supplementary Data online); however, \\nthese have not been consistently replicated.\\ndrug: clopidogrel\\nBackground.  Clopidogrel is a thienopyridine prodrug that \\nrequires hepatic biotransformation to form an active metabolite \\nthat selectively and irreversibly inhibits the purinergic P2RY12 \\nreceptor, and thereby platelet aggregation, for the platelet’s life span (~10 days). Only 15% of the prodrug is available for trans-\\nformation to the active agent; the other 85% is hydrolyzed by \\nesterases to inactive forms. Conversion of clopidogrel to its active metabolite requires two sequential oxidative steps involv-ing several CYP450 enzymes (e.g., CYP1A2, CYP2B6, CYP2C9, CYP2C19, and CYP3A4/5; Supplementary Figure S1 online). \\nA recent study suggested that paraoxonase 1 (PON1) may also \\nbe involved in clopidogrel activation.\\n14\\nLinking genetic variability to variability in drug-related phenotypes.  \\nClopidogrel is commonly prescribed for ACSs and/or follow-\\ning PCI.', start_char_idx=0, end_char_idx=3851, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=0.7613305330351511),\n",
       " NodeWithScore(node=TextNode(id_='e1562dd4-7f65-44b8-ac67-1c6ad0e4a1a3', embedding=None, metadata={'page_label': '1', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, excluded_embed_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], excluded_llm_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='ce9e11f0-55f3-4af7-ad73-ab77cecdd744', node_type=<ObjectType.DOCUMENT: '4'>, metadata={'page_label': '1', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='6cd0f707599c930dca2369fa1b1bf1c0579336017410bbd1113f065339c999d4'), <NodeRelationship.PREVIOUS: '2'>: RelatedNodeInfo(node_id='e9cb0c6e-d7dd-4dfe-9237-51320816eee6', node_type=<ObjectType.TEXT: '1'>, metadata={'page_label': '5', 'file_name': '21270794.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21270794.pdf', 'file_type': 'application/pdf', 'file_size': 158775, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='9de150ca514b9c4a776b5c10a6c52aa5e478f239774ed6ef36dcca9a405d1c9c'), <NodeRelationship.NEXT: '3'>: RelatedNodeInfo(node_id='0625cf87-e45e-442b-a300-8ac47159be9f', node_type=<ObjectType.TEXT: '1'>, metadata={}, hash='17cda3052ec13005cbe00f8b21a5291650871997d32f86679ae0fff5e1016264')}, text='Clini C al pharma C ology & Therapeu T i C s  1translation nature publishing group\\nCYP2C19 is one of the principal enzymes involved in the \\nbioactivation of the antiplatelet prodrug clopidogrel. A common loss-of-function allele, CYP2C19*2 (c.681G>A; \\nrs4244285), is associated with increased risk for serious adverse \\ncardiovascular events in both heterozygous and homozygous patients (~25–50% of the population) with acute coronary syndromes (ACSs) who are receiving clopidogrel, particularly among those undergoing percutaneous coronary intervention \\n(PCI). We provide evidence from published literature and guidelines for CYPC19 genotype–directed antiplatelet therapy \\n(periodically updated at http://www.pharmgkb.org).\\nThe purpose of these guidelines is to provide information \\nregarding the clinical use of CYP2C19  genotype test results \\nin patients requiring antiplatelet therapy. Detailed guidelines \\nfor use of other laboratory tests, such as platelet function monitoring, as well as a discussion of cost effectiveness, are \\nbeyond the scope of this article. The article does not focus on \\ndemographic and other clinical variables, such as adherence to \\ntherapy, age, diabetes mellitus, obesity, smoking, and concomi-\\ntant use of other drugs, that may influence clopidogrel efficacy \\nand clinical decision making. The Clinical Pharmacogenetics \\nImplementation Consortium of the National Institutes of \\nHealth’s Pharmacogenomics Research Network develops peer-\\nreviewed gene/drug guidelines that are published and updated periodically at http://www.pharmgkb.org on the basis of new \\ndevelopments in the field.1\\nFocused Literature r eview\\nA systematic literature review was conducted on CYP2C19  \\ngenotype and clopidogrel (see Supplementary Data  online). \\nGuidelines for antiplatelet therapy were developed based on interpretation of the literature by authors and experts in the \\nfield.\\nGene: CYP2C19\\nBackground.  The hepatic CYP2C19 enzyme contributes to the \\nmetabolism of many clinically relevant drugs such as antide -\\npressants, benzodiazepines, mephenytoin, some proton pump \\ninhibitors, and clopidogrel. Like many other CYP450  super -\\nfamily members, the CYP2C19 gene is highly polymorphic, hav -\\ning more than 25 known variant alleles (http://www.cypalleles.\\nki.se/cyp2c19.htm). The CYP2C19*1 allele is associated with \\nfunctional CYP2C19-mediated metabolism. The most common \\nCYP2C19 loss-of-function allele is *2 (c.681G>A; rs4244285), \\nwith allele frequencies of ~15% in Caucasians and Africans, and \\n29–35% in Asians (Supplementary Tables S3  and S4 online).', start_char_idx=0, end_char_idx=2579, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=0.7362817874252398),\n",
       " NodeWithScore(node=TextNode(id_='de5f93a0-ade7-4577-a281-5f2ebe5322e1', embedding=None, metadata={'page_label': '5', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, excluded_embed_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], excluded_llm_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='f12b7b29-16dd-43b1-88ad-f67f01cb2eff', node_type=<ObjectType.DOCUMENT: '4'>, metadata={'page_label': '5', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='967f0ffa24010c75c17232bddd5656ee85360bd216c098f5c6d719b75957ed69'), <NodeRelationship.PREVIOUS: '2'>: RelatedNodeInfo(node_id='ed111858-93f1-445f-bd86-85078c3e1959', node_type=<ObjectType.TEXT: '1'>, metadata={'page_label': '5', 'file_name': '21716271.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/21716271.pdf', 'file_type': 'application/pdf', 'file_size': 151974, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='8caed68a6a0a4745e93fa40058c21df07c563ec71d11e8c98834979056f4d28b'), <NodeRelationship.NEXT: '3'>: RelatedNodeInfo(node_id='732cc748-6a36-493b-a9c2-2ddfed75760a', node_type=<ObjectType.TEXT: '1'>, metadata={}, hash='fc5c8747a213bc1e2f73d9dfacbc8670e805d54224ad27b91a708a3573fca233')}, text='JAMA 302, \\n849–857 (2009).\\n8.\\n \\nFrére, C., Cuisset, T., Gaborit, B., Alessi, M.C. & Hulot, J.S. The CYP2C19*17 \\nallele is associated with better platelet response to clopidogrel in patients admitted for non-ST acute coronary syndrome. J. Thromb. Haemost. 7, \\n1409–1411 (2009).\\n9.\\n \\nSibbing, D. et al. Isolated and interactive impact of common CYP2C19 genetic \\nvariants on the antiplatelet effect of chronic clopidogrel therapy. J. Thromb. \\nHaemost. 8, 1685–1693 (2010).\\n10.\\n \\nTiroch, K.A. et al. Protective effect of the CYP2C19 *17 polymorphism with \\nincreased activation of clopidogrel on cardiovascular events. Am. Heart J. 160, \\n506–512 (2010).\\n11.\\n \\nSibbing, D. et al. Cytochrome 2C19*17 allelic variant, platelet aggregation, \\nbleeding events, and stent thrombosis in clopidogrel-treated patients with \\ncoronary stent placement. Circulation 121, 512–518 (2010).12.\\n \\nGeisler, T. et al. CYP2C19 and nongenetic factors predict poor responsiveness \\nto clopidogrel loading dose after coronary stent implantation. \\nPharmacogenomics 9, 1251–1259 (2008).\\n13.\\n \\nSimon, T. et al.; French Registry of Acute ST-Elevation and Non-ST-Elevation \\nMyocardial Infarction (FAST-MI) Investigators. Genetic determinants of response to clopidogrel and cardiovascular events. N. Engl. J. Med. 360, \\n363–375 (2009).\\n14.\\n \\nBouman, H.J. et al. Paraoxonase-1 is a major determinant of clopidogrel \\nefficacy. Nat. Med. 17, 110–116 (2011).\\n15.\\n \\nKim, K.A., Park, P .W., Hong, S.J. & Park, J.Y. The effect of CYP2C19 polymorphism \\non the pharmacokinetics and pharmacodynamics of clopidogrel: a possible mechanism for clopidogrel resistance. Clin. Pharmacol. Ther. 84, 236–242 \\n(2008).\\n16.\\n \\nGiusti, B. et al. Relation of cytochrome P450 2C19 loss-of-function \\npolymorphism to occurrence of drug-eluting coronary stent thrombosis. Am. \\nJ. Cardiol. 103, 806–811 (2009).\\n17.\\n \\nSibbing, D. et al. Cytochrome P450 2C19 loss-of-function polymorphism and \\nstent thrombosis following percutaneous coronary intervention. Eur. Heart J.  \\n30, 916–922 (2009).\\n18.\\n \\nMega, J.L. et al. Reduced-function CYP2C19 genotype and risk of adverse \\nclinical outcomes among patients treated with clopidogrel predominantly for PCI: a meta-analysis. JAMA 304, 1821–1830 (2010).\\n19.\\n \\nParé, G. et al. Effects of CYP2C19 genotype on outcomes of clopidogrel \\ntreatment. N. Engl. J. Med. 363, 1704–1714 (2010).\\n20.\\n \\nWallentin, L. et al.; PLATO investigators. Effect of CYP2C19 and ABCB1 single \\nnucleotide polymorphisms on outcomes of treatment with ticagrelor versus \\nclopidogrel for acute coronary syndromes: a genetic substudy of the PLATO trial. Lancet 376, 1320–1328 (2010).\\n21. \\n \\nHolmes, D.R.  Jr, Dehmer, G.J., Kaul, S., Leifer, D., O’Gara, P .T. & Stein, C.M.', start_char_idx=2768, end_char_idx=5485, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=0.7312752016477744),\n",
       " NodeWithScore(node=TextNode(id_='f94eb2ff-5c09-4065-b1b0-aa378b0d878c', embedding=None, metadata={'page_label': '2', 'file_name': '35034351.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/35034351.pdf', 'file_type': 'application/pdf', 'file_size': 176977, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, excluded_embed_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], excluded_llm_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='96050a46-c7ed-4545-8e5c-eff6d63d368f', node_type=<ObjectType.DOCUMENT: '4'>, metadata={'page_label': '2', 'file_name': '35034351.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/35034351.pdf', 'file_type': 'application/pdf', 'file_size': 176977, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='bf1638951e2349075a18af5d19ca1519b7903bdc04beb9dd0535205e9f00ea0f'), <NodeRelationship.PREVIOUS: '2'>: RelatedNodeInfo(node_id='e11d3b52-7502-4230-8590-dbb4df1ce0fe', node_type=<ObjectType.TEXT: '1'>, metadata={'page_label': '2', 'file_name': '35034351.pdf', 'file_path': '/home/karlkeat/projects/PGx-LLM-RAG/cpic_guidelines/35034351.pdf', 'file_type': 'application/pdf', 'file_size': 176977, 'creation_date': '2024-04-02', 'last_modified_date': '2024-04-02'}, hash='6447a1c449ecd20d880fd341b5d213031dcf0e9e8cdcb2feaebd4f5f5c486006'), <NodeRelationship.NEXT: '3'>: RelatedNodeInfo(node_id='793f581c-8f71-4e2f-82d2-76f46fe7ea1e', node_type=<ObjectType.TEXT: '1'>, metadata={}, hash='9e44e4c3c1591e8e012d0c608183ace60ff3a63eb20627d27f5a3f6ab411bd11')}, text='Incidental findings\\nNo inherited diseases or conditions have been consistently or \\nstrongly linked to germline genetic variants in CYP2C19  inde -\\npendent of drug metabolism and response.\\nOther considerations\\nCYP2C19  genetic variation does not account for all of the vari -\\nability in clopidogrel response. Some studies have implicated \\nvariants in other genes associated with clopidogrel response \\n(e.g., ABCB1, B4GALT2 , CES1 , CYP2B6 , CYP2C9 , P2RY12 , \\nand PON1 ); however, these have not been consistently repli -\\ncated, and CYP2C19  is the most validated genetic determinant \\nof clopidogrel response.11– 13 Consequently, this guideline on \\ngenotype-  guided antiplatelet therapy is limited to CYP2C19  \\nvariant alleles.\\nTable 1 Assignment of predicted CYP2C19 phenotype based on genotype\\nPredicted phenotype GenotypeExamples of CYP2C19  \\ndiplotypesa\\nCYP2C19 ultrarapid metabolizer An individual carrying two increased function alleles *17/*17\\nCYP2C19 rapid metabolizer An individual carrying one normal function allele and \\none increased function allele*1/*17\\nCYP2C19 normal metabolizer An individual carrying two normal function alleles *1/*1\\nCYP2C19 likely intermediate metabolizerbAn individual carrying one normal function allele \\nand one decreased function allele or one increased \\nfunction allele and one decreased function allele or \\ntwo decreased function alleles*1/*9 , *9/*17 , *9/*9\\nCYP2C19 intermediate metabolizer An individual carrying one normal function allele \\nand one no function allele or one increased function \\nallele and one no function allele*1/*2 , *1/*3 , *2/*17 , *3/*17\\nCYP2C19 likely poor metabolizerbAn individual carrying one decreased function allele \\nand one no function allele*2/*9 , *3/*9\\nCYP2C19 poor metabolizer An individual carrying two no function alleles *2/*2 , *3/*3 , *2/*3\\nIndeterminate metabolizer An individual carrying one or two uncertain function \\nalleles*1/*12 , *2/*12 , *12/*14\\naPlease refer to the CYP2C19  Diplotype-  Phenotype Table online for a complete list. For allele functions and population-  specific allele and phenotype frequencies, \\nplease refer to the CYP2C19  Allele Functionality Table and the CYP2C19  Allele Frequency Table online.8,9\\n bThere are limited data to characterize the function of decreased function alleles.\\nCPIC UPDATE', start_char_idx=3029, end_char_idx=5338, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=0.7309705152645259)]"
      ]
     },
     "execution_count": 404,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpic_guideline_retriever = cpic_guideline_index.as_retriever(similarity_top_k=5)\n",
    "cpic_guideline_nodes = cpic_guideline_retriever.retrieve(\"CYP2C19*1/*2, clopidogrel?\")\n",
    "cpic_guideline_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cpic_guidelines(query_str, top_k=5):\n",
    "    cpic_guideline_retriever = cpic_guideline_index.as_retriever(similarity_top_k=top_k)\n",
    "    return cpic_guideline_retriever.retrieve(query_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CPIC Database Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [],
   "source": [
    "from llama_index.core import SQLDatabase\n",
    "from sqlalchemy import (\n",
    "    create_engine,\n",
    "    MetaData,\n",
    "    text\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gene',\n",
       " 'pair',\n",
       " 'drug',\n",
       " 'guideline',\n",
       " 'test_alert',\n",
       " 'allele_definition',\n",
       " 'allele_frequency',\n",
       " 'allele',\n",
       " 'population',\n",
       " 'publication',\n",
       " 'change_log',\n",
       " 'recommendation',\n",
       " 'gene_result',\n",
       " 'gene_result_lookup',\n",
       " 'gene_result_diplotype',\n",
       " 'file_artifact',\n",
       " 'flyway_schema_history',\n",
       " 'file_note',\n",
       " 'sequence_location',\n",
       " 'term',\n",
       " 'statistic',\n",
       " 'publication_supplement',\n",
       " 'allele_location_value',\n",
       " 'file_artifact_history']"
      ]
     },
     "execution_count": 356,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "engine = create_engine(\"postgresql://karlkeat@localhost:5432/cpic_db_rag\", # Replace with your own db path\n",
    "                       connect_args={'options': '-csearch_path=cpic'})\n",
    "metadata_obj = MetaData()\n",
    "metadata_obj.create_all(engine)\n",
    "metadata_obj.reflect(engine)\n",
    "table_names = list(metadata_obj.tables.keys())\n",
    "table_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [],
   "source": [
    "sql_database = SQLDatabase(engine)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Database query test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('CYP2C19', '*2', 'rs12769205'), ('CYP2C19', '*2', 'rs58973490'), ('CYP2C19', '*2', 'rs4244285'), ('CYP2C19', '*2', 'rs3758581')]\n"
     ]
    }
   ],
   "source": [
    "stmt = \"select d.genesymbol, d.name, sl.dbsnpid\\nfrom allele_definition d\\njoin allele_location_value alv on d.id = alv.alleledefinitionid\\njoin sequence_location sl on alv.locationid = sl.id\\nwhere d.genesymbol='CYP2C19' and d.name='*2';\"\n",
    "with engine.connect() as connection:\n",
    "    results = connection.execute(text(stmt)).fetchall()\n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Text-to-SQL Query Engine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n"
     ]
    }
   ],
   "source": [
    "from llama_index.core.objects import (\n",
    "    SQLTableNodeMapping,\n",
    "    ObjectIndex,\n",
    "    SQLTableSchema,\n",
    ")\n",
    "from llama_index.core import VectorStoreIndex\n",
    "from llama_index.core.retrievers import NLSQLRetriever\n",
    "\n",
    "# set Logging to DEBUG for more detailed outputs\n",
    "table_node_mapping = SQLTableNodeMapping(sql_database)\n",
    "table_schema_objs = [\n",
    "    (SQLTableSchema(table_name=table)) for table in table_names\n",
    "]  # add a SQLTableSchema for each table\n",
    "obj_index = ObjectIndex.from_objects(\n",
    "    table_schema_objs,\n",
    "    table_node_mapping,\n",
    "    VectorStoreIndex\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene': \"- table: `gene`\\n\\nThis model represents a human gene and is in the table `gene`. It includes properties that are specific to the gene \\nlike references to other data sources such as NCBI, HGNC, and Ensembl.\\n\\nThe primary key for a Gene is the `symbol` property. This is the official HGNC symbol for that gene.\\n\\nThe presence of a gene in this model does NOT guarantee that a CPIC guideline exists for that gene.\\n\\nAs an example, here's how you would get information about CYP2D6.\\n\\n\\nselect * from gene where symbol='CYP2D6';\",\n",
       " 'drug': '- table: `drug`\\n\\nThe \"drug\" model represents a drug referenced somewhere in the data model. It includes references to other drug data \\nsources like ATC, RxNorm, and UMLS.\\n\\nThe primary key for a Drug is the `drugid` property. The value of this property is in the form \"source:id\". For \\nexample: `RxNorm:2670` for codeine. We attempt to use RxNorm as the primary key for most drugs, but some CPIC drug \\nentities don\\'t have equivalent values in RxNorm. Those entities will use some other resource. \\n\\nThe `flowChart` property of the Drug model is a URL where the flowchart diagram image can be found.\\n\\nThe presence of a drug in this model does NOT guarantee that a CPIC guideline exists for that drug. \\n\\nAs an example, here\\'s how you would get information about codeine.\\n\\n\\nselect * from drug where name=\\'codeine\\';',\n",
       " 'allele': '- tables: `allele`, `allele_definition` \\n\\nThis model represents a named allele of a Gene (e.g. \"*2\"). It is split into two tables, the `allele` table and the `allele_definition`\\ntable. We split this model to avoid duplicating definition data for alleles that differ only on copy number.\\n\\nFor example, the alleles \"CYP2D6\\\\*2\", \"CYP2D6\\\\*2x2\", and \"CYP2D6\\\\*2≥3\" are all defined by the same SNPs but differ in the \\ncopy number of the CYP2D6 Gene. Those will be 3 different rows in the `allele` table that all refer to the same row in \\nthe `allele_definition` table.\\n\\nThe `allele_definition` table links to the `allele_locaton_value` table. This contains the specific variant alleles\\nused to define the named allele and a reference to the locations that allele appears on sequences. The table that holds \\nthe location data is called `sequence_location`.\\n\\nHere\\'s how to get a list of all alleles for CYP2D6 including copy number alleles.\\n\\n\\nselect * from allele where genesymbol=\\'CYP2D6\\';\\n\\n\\nTo get the definition data for CYP2D6\\\\*3 you must join the `allele_definition` to the table `allele_location_value` \\nwhich specifies the variants used to define that allele and then join that to `sequence_location` to get information on \\nthe locations (like dbSNP ID). \\n\\n\\nselect d.genesymbol, d.name, sl.dbsnpid, alv.variantallele\\nfrom allele_definition d\\njoin allele_location_value alv on d.id = alv.alleledefinitionid\\njoin sequence_location sl on alv.locationid = sl.id\\nwhere d.genesymbol=\\'CYP2D6\\' and d.name=\\'*3\\';',\n",
       " 'guideline': '- table: `guideline`\\n\\nThis model represents a logical \"guideline\" published by CPIC. This data is in the `guideline` table. A single \\n\"guideline\" in this model encompasses the original publication of the guideline plus any subsequent update publications. \\n\\nTo get a list of all CPIC guidelines:\\n\\n\\nselect * from guideline;\\n  \\n\\nAn example of querying the API for the thiopurines guideline and all publications for that guideline:',\n",
       " 'publication': '- table: `publication`\\n\\nThe publication model represents published literature, typically with a reference to PubMed. This table will include information \\nabout publications used in different contexts in the CPIC database. This table is _not_ exclusively publications by CPIC or \\nCPIC curators. It will include guideline publications, but it will also include information about publications used as \\nreferences when assigning allele function, for example.\\n\\nTo list all guideline publications:\\n\\n\\nselect * from publication where guidelineid is not null;',\n",
       " 'recommendation': '- tables: `recommendation`\\n\\nThe recommendation table contains implications, recommendations, comments, and other information published in CPIC \\nguidelines (typically, \"Table 2\" of the guideline publication). The rows in this table will reference the guideline \\nthat they originate from with the `guidelineid` property. The recommendations will also reference the specific drug they \\napply to with the `drugid` column. Some guidelines may have different recommendations for people in different \\npopulations so a `population` column exists to differentiate those. Populations can indicate age groups like pediatric \\npopulation or people who are on particular drug therapies.\\n\\nGenerally, to get a guaranteed-unique recommendation the lookup query must specify:\\n\\n- drug\\n- population\\n- gene lookup key (described below)\\n\\nRecommendations use one or more genes to give specific information. Different genes can use different methods for \\nmatching to recommendations. The currently supported methods for gene-recommendation matches are by phenotype, by \\nactivity score, or by allele status. You can find the lookup method for a gene by reading the `lookupmethod` column of \\nthe \\'gene\\' table.\\n\\nPhenotype lookup maps a pair of allele functions to a metabolizer phenotype (e.g. \"Poor Metabolizer\"). This is \\nrepresented in the `phenotypes` column. Activity score maps pairs of allele functions to numerical descriptions instead \\nof a \"metabolizer\" text. This is represented in the `activityScore` column. Allele status indicates the presence of \\nparticular alleles in a gene and does not use any indication of allele function or metabolizer status. This is indicated in \\nthe `allele_status` column.   \\n\\nThis leads to a problem for guidelines that use more than one gene which have different lookup methods. You would have \\nto look in different columns for different guidelines in order to match known gene statuses to recommendations. To make \\nthis use case slightly easier, we added a `lookupKey` column which combines all gene statuses regardless of whether \\nthey use phenotype, activity score, or allele status. This gives one value that can consistently be used for \\nrecommendation lookups.\\n\\nAs an example, here\\'s how you would find the recommendation for voriconazole in the pediatrics population with a \\n\"Normal Metabolizer\" phenotype.\\n\\n\\nselect * from recommendation where lookupKey=\\'{\"CYP2C19\": \"Poor Metabolizer\"}\\' and drugid=\\'RxNorm:121243\\' and population=\\'pediatrics\\';',\n",
       " 'test_alert': '- table: `test_alert`\\n\\nThe test alert table has example text for specific gene activity scores or phenotypes for a specific drug. Just like the \\nrecommendation table in the previous section, this table also includes population-specific rows. Again, the specific \\nalerts should be queried by phenotype or activity score depending on what the `lookupMethod` in the `gene` table \\nspecifies for the gene (or genes) used in the test alert.\\n\\nThe example for `test_alert` can use the same options as the previous recommendation example.\\n\\n\\nselect * from test_alert where phenotype=\\'{\"CYP2C19\": \"Poor Metabolizer\"}\\' and drugid=\\'RxNorm:121243\\' and population=\\'pediatrics\\';',\n",
       " 'gene_results_and_diplotypes': '- tables: `gene_result` and `diplotype`\\n\\nGene result information exists in the `gene_result`, `gene_result_lookup`, and `gene_result_diplotype` tables. The \\nthree tables represent the three levels of specificity for result information. A gene result (e.g. \"Poor Metabolizer\") \\nfound in the `gene_result` table can be applicable to multiple function combinations or activity scores.\\n\\nThe `gene_result` table includes EHR priority and example consultation text, if it has been part of a published \\nguideline.\\n\\nThe `gene_result_lookup` table tracks function combinations and activity scores that are assigned to \\nthe result in `gene_result`. For example, a row with two \"No Function\" alleles in the `gene_result_lookup` table \\nmay map to a \"Poor Metabolizer\" row in the `gene_result` table.\\n\\nThe `gene_result_diplotype` table takes it even further and maps the function pairs to specific diplotypes. For example, a \\nrow for \"CYP2C19 \\\\*2/\\\\*4\" in the `gene_result_diplotype` table will map to two \"No Function\" alleles in the \\n`gene_result_lookup` table which will then map to a \"Poor Metabolizer\" phenotype in the `gene_result` table.\\n\\nFor example, to see all possible phenotypes for CYP2C19\\n\\n\\nselect * from gene_result g where genesymbol=\\'CYP2C19\\';\\n\\n\\nAn example of looking up the result by activity score instead of by function. This query finds the CYP2C19 phenotype \\nwhen one allele gets a score of \"0\" and one allele gets a score of \"0.5\".  \\n\\n\\nselect g.genesymbol, g.result, pf.*\\nfrom gene_result g\\n    join gene_result_lookup pf on g.id = pf.phenotypeid\\nwhere genesymbol=\\'CYP2C9\\' and lookupkey=\\'{\"0\": 1, \"0.5\": 1}\\';\\n\\n\\nThe data that comes from joining these three tables together can be tricky to express via the API. The `diplotype` \\nview helps solve this problem. It\\'s a pre-joined view of all three tables with a subset of the most relevant columns. \\nYou can query by gene and diplotype to get the function information and result information. This can be especially \\nuseful when you have a diplotype and want to look up recommendation data which is keyed by result.\\n\\nFor example, here\\'s how you would query for data on UGT1A1 \\\\*27/\\\\*80.',\n",
       " 'pair': '- tables: `pair_view` and `pair`\\n\\nThe `pair` table lists gene-drug pairs that are tracked by CPIC. Each pair is assigned a CPIC level and tracks \\ninformation about FDA drug label testing recommendations and PharmGKB clinical annotation levels. Other properties of \\nthe pair include the top PharmGKB clinical annotation level and the FDA PGx testing level annotated by PharmGKB. These \\npairs can also optionally be related to guidelines.\\n\\nThe `pair` table contains the raw data, but the `pair_view` view also includes helpful data like the drug name and the \\nguideline name. Most people will want to use the `pair_view` since it saves the work of joining to other tables.\\n\\nFor example, here\\'s how to query all \"A\"-level pairs.\\n\\n\\nselect * from pair_view p where cpiclevel=\\'A\\';',\n",
       " 'allele_frequency': '- tables: `population_frequency_view`\\n\\nAllele frequency data is mainly in the `allele_frequency` table. This table has a foreign key reference off to the \\npreviously mentioned `allele` table and another foreign key reference to a `population` table. Each row in \\n`allele_frequency` is unique to an allele and population combination. The population data contains a descriptive name \\nfor the population (column `population`) but also a higher-level grouping (column `ethnicity`). The rows in `population` \\nare unique to a given `publication` but they may share population and ethnicity descriptions with other publications.\\n\\nThe `allele_frequency` table stores frequency data in two columns `frequency` and `label`. The `frequency` column is the \\nnumerical representation of the frequency and suitable for aggregation or mathematical analysis. The `label` column is \\nthe way the source document specified the frequency and, thus, is suitable for reporting for human legibility.\\n\\nThe `allele_frequency` table will not have data on reference alleles (e.g. `*1`) since studies, typically, do not \\ndirectly test for it.\\n\\nThere is also a view called `population_frequency_view`. This view joins the `allele`, `allele_frequency`, and \\n`population` tables while doing some summary statistics like \"weighted average\". This view groups the populations into \\ntheir `population_group` values which is typical for most reporting.\\n\\nAs an example, here\\'s how you could use the `population_frequency_view` to find all allele-population combinations for \\nCYP2C9 that have an average weighted frequency greater than 0.01.  \\n\\n\\nselect * from population_frequency_view where genesymbol=\\'CYP2C9\\' and freq_weighted_avg>0.01;'}"
      ]
     },
     "execution_count": 369,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"./cpic_db_docs/cpic_wiki.md\") as infile:\n",
    "    text = infile.read()\n",
    "    \n",
    "cpic_db_wiki = {}\n",
    "for section in text.split(\"###\")[1:]:\n",
    "    header = section.split(\"\\n\")[0].strip().lower().replace(\" \", \"_\")\n",
    "    body = \"\\n\".join(section.split(\"\\n\")[2:]).strip()\n",
    "    cpic_db_wiki[header] = body\n",
    "cpic_db_wiki"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_text_to_sql(text_to_sql_query):\n",
    "    # Fetch relevant SQL table documentation\n",
    "    table_retriever = obj_index.as_retriever(similarity_top_k=5)\n",
    "    tables_to_query = [x.table_name for x in table_retriever.retrieve(text_to_sql_query)]\n",
    "\n",
    "    context = \"\"\n",
    "\n",
    "    # Add wiki docs to context\n",
    "    for table in tables_to_query:\n",
    "        if table in cpic_db_wiki:\n",
    "            context += cpic_db_wiki[table]\n",
    "\n",
    "    if len(context) > 0:\n",
    "        context = \"Here is some useful information on the SQL table structure: \\n\" + context + \"\\n\"\n",
    "\n",
    "    # Query text2sql model\n",
    "    nl_sql_retriever = NLSQLRetriever(\n",
    "        sql_database, return_raw=True, tables = tables_to_query, handle_sql_errors=True\n",
    "    )\n",
    "    return nl_sql_retriever.retrieve_with_metadata(context + text_to_sql_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:llama_index.core.indices.struct_store.sql_retriever:> Table desc str: Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "> Table desc str: Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "> Table desc str: Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "> Table desc str: Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([NodeWithScore(node=TextNode(id_='17585b84-39ab-4a4f-be99-d78db6cec36f', embedding=None, metadata={}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text=\"[('CYP2C19', '*2', 'rs12769205'), ('CYP2C19', '*2', 'rs58973490'), ('CYP2C19', '*2', 'rs4244285'), ('CYP2C19', '*2', 'rs3758581')]\", start_char_idx=None, end_char_idx=None, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=None)],\n",
       " {'sql_query': \"select d.genesymbol, d.name, sl.dbsnpid\\nfrom allele_definition d\\njoin allele_location_value alv on d.id = alv.alleledefinitionid\\njoin sequence_location sl on alv.locationid = sl.id\\nwhere d.genesymbol='CYP2C19' and d.name='*2';\",\n",
       "  'result': [('CYP2C19', '*2', 'rs12769205'),\n",
       "   ('CYP2C19', '*2', 'rs58973490'),\n",
       "   ('CYP2C19', '*2', 'rs4244285'),\n",
       "   ('CYP2C19', '*2', 'rs3758581')],\n",
       "  'col_keys': ['genesymbol', 'name', 'dbsnpid']})"
      ]
     },
     "execution_count": 362,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = run_text_to_sql(\"What are the dbsnpids (not pharmvarIDs) associated with CYP2C19 *2?\")\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:llama_index.core.indices.struct_store.sql_retriever:> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'gene_result_diplotype' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', functionphenotypeid (INTEGER): 'An ID referencing a gene_result_lookup record, required', diplotype (TEXT): 'A diplotype for the gene in the form Allele1/Allele2, required', diplotypekey (JSONB): 'A normalized JSON version of the diplotype for use in lookups. Should be an object with the allele names as properties and the counts as the values. Required', frequency (JSONB), with comment: (Specific diplotypes that are associated with a gene result. This table is a child of gene_result_lookup and distantly of gene_result) and foreign keys: ['functionphenotypeid'] -> gene_result_lookup.['id'].\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([NodeWithScore(node=TextNode(id_='28553aaa-6340-4269-9f31-fd48c1cfef1d', embedding=None, metadata={}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text='[(None,)]', start_char_idx=None, end_char_idx=None, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=None)],\n",
       " {'sql_query': \"SELECT a.functionalstatus\\nFROM allele a\\nJOIN allele_definition ad ON a.definitionid = ad.id\\nWHERE ad.genesymbol = 'CYP2C19' AND ad.name = '*3';\",\n",
       "  'result': [(None,)],\n",
       "  'col_keys': ['functionalstatus']})"
      ]
     },
     "execution_count": 363,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = run_text_to_sql(\"What is the clinical functionality of the CYP2C19*3 allele?\")\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:llama_index.core.indices.struct_store.sql_retriever:> Table desc str: Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'population' has columns: id (INTEGER): 'A synthetic primary key ID', publicationid (INTEGER): 'A reference to a publication', ethnicity (TEXT): 'The major ethnicity grouping of this population', population (TEXT): 'The descriptive name of this population, optional', populationinfo (TEXT): 'Further information about this population, optional', subjecttype (TEXT): 'Information about the types of subjects in this population', subjectcount (INTEGER): 'The total number of subjects in this population', version (INTEGER): 'The version number, iterates on modification', with comment: (A population description of a group of subjects) and foreign keys: ['publicationid'] -> publication.['id'].\n",
      "> Table desc str: Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'population' has columns: id (INTEGER): 'A synthetic primary key ID', publicationid (INTEGER): 'A reference to a publication', ethnicity (TEXT): 'The major ethnicity grouping of this population', population (TEXT): 'The descriptive name of this population, optional', populationinfo (TEXT): 'Further information about this population, optional', subjecttype (TEXT): 'Information about the types of subjects in this population', subjectcount (INTEGER): 'The total number of subjects in this population', version (INTEGER): 'The version number, iterates on modification', with comment: (A population description of a group of subjects) and foreign keys: ['publicationid'] -> publication.['id'].\n",
      "> Table desc str: Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'population' has columns: id (INTEGER): 'A synthetic primary key ID', publicationid (INTEGER): 'A reference to a publication', ethnicity (TEXT): 'The major ethnicity grouping of this population', population (TEXT): 'The descriptive name of this population, optional', populationinfo (TEXT): 'Further information about this population, optional', subjecttype (TEXT): 'Information about the types of subjects in this population', subjectcount (INTEGER): 'The total number of subjects in this population', version (INTEGER): 'The version number, iterates on modification', with comment: (A population description of a group of subjects) and foreign keys: ['publicationid'] -> publication.['id'].\n",
      "> Table desc str: Table 'allele_frequency' has columns: alleleid (INTEGER): 'The ID of the allele for this observation', population (INTEGER): 'The ID of the population for this observation', frequency (NUMERIC): 'The numeric representation of this frequency', label (TEXT): 'The textual label for this frequency, \"-\" means no observation made', version (INTEGER): 'The version number, iterates on modification', with comment: (A frequency observation for a particular allele in a particular population) and foreign keys: ['alleleid'] -> allele.['id'], ['population'] -> population.['id'].\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'population' has columns: id (INTEGER): 'A synthetic primary key ID', publicationid (INTEGER): 'A reference to a publication', ethnicity (TEXT): 'The major ethnicity grouping of this population', population (TEXT): 'The descriptive name of this population, optional', populationinfo (TEXT): 'Further information about this population, optional', subjecttype (TEXT): 'Information about the types of subjects in this population', subjectcount (INTEGER): 'The total number of subjects in this population', version (INTEGER): 'The version number, iterates on modification', with comment: (A population description of a group of subjects) and foreign keys: ['publicationid'] -> publication.['id'].\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([NodeWithScore(node=TextNode(id_='5814f13f-23ee-40b5-a749-9d4b75a2de58', embedding=None, metadata={}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text=\"[(Decimal('0.14685704098527318700'),)]\", start_char_idx=None, end_char_idx=None, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=None)],\n",
       " {'sql_query': \"SELECT freq_weighted_avg \\nFROM population_frequency_view \\nWHERE genesymbol = 'CYP2C19' \\nAND population_group = 'European' \\nAND name = '*2';\",\n",
       "  'result': [(Decimal('0.14685704098527318700'),)],\n",
       "  'col_keys': ['freq_weighted_avg']})"
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = run_text_to_sql(\"What is the frequency weighted average of CYP2C19*2 in the European group?\")\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:llama_index.core.indices.struct_store.sql_retriever:> Table desc str: Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'guideline' has columns: id (INTEGER): 'A synthetic numerical ID, primary key', version (INTEGER): 'The version number, iterates on modification', name (TEXT): 'The name (title) of this guideline, required', url (TEXT): 'The URL of this guideline on the cpicpgx.org domain, optional', pharmgkbid (ARRAY): 'The IDs from PharmGKB for their annotations of this guideline, optional', genes (ARRAY): 'The subject genes of this guideline', notesonusage (TEXT): 'Notes about using a particular guideline within this system. This may include. You may find information here when data for a guideline seems missing or incomplete.', with comment: (A guideline for a drug or group of drugs) and foreign keys: .\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "> Table desc str: Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'guideline' has columns: id (INTEGER): 'A synthetic numerical ID, primary key', version (INTEGER): 'The version number, iterates on modification', name (TEXT): 'The name (title) of this guideline, required', url (TEXT): 'The URL of this guideline on the cpicpgx.org domain, optional', pharmgkbid (ARRAY): 'The IDs from PharmGKB for their annotations of this guideline, optional', genes (ARRAY): 'The subject genes of this guideline', notesonusage (TEXT): 'Notes about using a particular guideline within this system. This may include. You may find information here when data for a guideline seems missing or incomplete.', with comment: (A guideline for a drug or group of drugs) and foreign keys: .\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "> Table desc str: Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'guideline' has columns: id (INTEGER): 'A synthetic numerical ID, primary key', version (INTEGER): 'The version number, iterates on modification', name (TEXT): 'The name (title) of this guideline, required', url (TEXT): 'The URL of this guideline on the cpicpgx.org domain, optional', pharmgkbid (ARRAY): 'The IDs from PharmGKB for their annotations of this guideline, optional', genes (ARRAY): 'The subject genes of this guideline', notesonusage (TEXT): 'Notes about using a particular guideline within this system. This may include. You may find information here when data for a guideline seems missing or incomplete.', with comment: (A guideline for a drug or group of drugs) and foreign keys: .\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "> Table desc str: Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'recommendation' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', guidelineid (INTEGER): 'The guideline this recommendation appears in', drugid (VARCHAR(20)): 'The drug this recommendation is for', implications (JSONB): 'Implications for phenotypic measures, this is a JSON mapping of gene to implication', drugrecommendation (TEXT): 'Dosing or therapeutic recommendations, depending on particular drug', classification (VARCHAR(20)): 'Classification of recommendations, described in supplementary meterial', phenotypes (JSONB): 'Phenotypes that this recommendation applies to, this is a JSON mapping of gene to phenotype', activityscore (JSONB): 'Activity score that this recommendation applies to, this is a JSON mapping of gene to score value', allelestatus (JSONB): 'Whether or not an allele is present, used mainly for HLA genes, and used for recommendation lookups. This is a JSON mapping of gene to allele status (positive/negative)', lookupkey (JSONB): 'A key to use for finding a specific recommendation. Made of a JSON object of gene symbol to key value. The key value can be one of phenotype, activity score, or allele status depending on the gene.', population (TEXT): 'The population this recommendation is applicable to', comments (TEXT): 'Optional comments about the recommendation', version (INTEGER), dosinginformation (BOOLEAN): 'The guideline recommends a dose adjustment based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', alternatedrugavailable (BOOLEAN): 'The guideline recommends that an alternate drug be selected for patients based on genetic variants or phenotypes (e.g. \"poor metabolizers\") as a primary course of action. However, the guideline may also describe extenuating circumstances where this action is not appropriate.', otherprescribingguidance (BOOLEAN): 'The guideline and other prescribing guidance not described as either dose adjustment or alternate drug.', with comment: (Recommendations for a gene result pulled from a guideline) and foreign keys: ['drugid'] -> drug.['drugid'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'guideline' has columns: id (INTEGER): 'A synthetic numerical ID, primary key', version (INTEGER): 'The version number, iterates on modification', name (TEXT): 'The name (title) of this guideline, required', url (TEXT): 'The URL of this guideline on the cpicpgx.org domain, optional', pharmgkbid (ARRAY): 'The IDs from PharmGKB for their annotations of this guideline, optional', genes (ARRAY): 'The subject genes of this guideline', notesonusage (TEXT): 'Notes about using a particular guideline within this system. This may include. You may find information here when data for a guideline seems missing or incomplete.', with comment: (A guideline for a drug or group of drugs) and foreign keys: .\n",
      "\n",
      "Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([NodeWithScore(node=TextNode(id_='71ede359-66ca-4fee-9703-7372177889e8', embedding=None, metadata={}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, text='[]', start_char_idx=None, end_char_idx=None, text_template='{metadata_str}\\n\\n{content}', metadata_template='{key}: {value}', metadata_seperator='\\n'), score=None)],\n",
       " {'sql_query': 'select * from recommendation where lookupKey=\\'{\"CYP2C19\": \"*2/*2\"}\\' and drugid=\\'RxNorm:1205\\';',\n",
       "  'result': [],\n",
       "  'col_keys': ['id',\n",
       "   'guidelineid',\n",
       "   'drugid',\n",
       "   'implications',\n",
       "   'drugrecommendation',\n",
       "   'classification',\n",
       "   'phenotypes',\n",
       "   'activityscore',\n",
       "   'allelestatus',\n",
       "   'lookupkey',\n",
       "   'population',\n",
       "   'comments',\n",
       "   'version',\n",
       "   'dosinginformation',\n",
       "   'alternatedrugavailable',\n",
       "   'otherprescribingguidance']})"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = run_text_to_sql(\"Should someone with CYP2C19*2/*2 take clopidogrel?\")\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Joint SQL + Guidelines LLM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "metadata": {},
   "outputs": [],
   "source": [
    "def joint_query(query_str):\n",
    "    # Call CPIC vector DB model to retreive relevant guideline snippets\n",
    "    guidelines = get_cpic_guidelines(query_str, top_k=3)\n",
    "    guidelines = \"\\n\".join([x.node.text for x in guidelines])\n",
    "\n",
    "    # Call text2sql model to retreive relevant DB output\n",
    "    sql_results = run_text_to_sql(query_str)[1]\n",
    "    if \"result\" in sql_results:\n",
    "        sql_results = sql_results[\"result\"]\n",
    "    else:\n",
    "        sql_results = \"\"\n",
    "\n",
    "    # Attach context to query and submit\n",
    "    prompt = f\"Here is some useful context to answer this query, do not use any outside information: {guidelines}\\n{sql_results}\\nHere is the query: {query_str}\"\n",
    "    return llm.complete(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:llama_index.core.indices.struct_store.sql_retriever:> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'gene' has columns: symbol (VARCHAR(20)): 'Approved HGNC symbol, primary key.', chr (VARCHAR(20)): 'Chromosome symbol. In the form chr##, where ## is the number or X/Y.', genesequenceid (TEXT): 'The RefSeq ID for the sequence that represents this gene, starts with \"NG_\". No version suffix.', proteinsequenceid (TEXT): 'The RefSeq ID for the sequence that represents the protein product of this gene, starts with \"NP_\". No version suffix.', chromosequenceid (TEXT): 'The RefSeq ID for the sequence that represents the chromosome this gene is on, starts with \"NC_\". No version suffix.', mrnasequenceid (TEXT): 'The RefSeq ID for the sequence that represents the translation of this gene, starts with \"NM_\". No version suffix.', hgncid (TEXT): 'The HGNC numerical ID number for this gene prefixed by \"HGNC:\"', ncbiid (TEXT): 'The NCBI Gene (Entrez) ID number for this gene', ensemblid (TEXT): 'The Ensembl ID for this gene', pharmgkbid (TEXT): 'The ID for this gene in PharmGKB.', frequencymethods (TEXT): 'Text documentation of the methods and caveats for allele frequency data', lookupmethod (TEXT): 'The way to lookup information about diplotypes of this gene, should use ACTIVITY_SCORE, PHENOTYPE, or ALLELE_STATUS', version (INTEGER), notesondiplotype (TEXT): 'Optional notes about the diplotypes of the gene', url (TEXT), functionmethods (TEXT), notesonallelenaming (TEXT), includephenotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display phenotype-level frequencies. a true value does not imply existence frequency data', includediplotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display diplotype-level frequencies. a true value does not imply existence frequency data', with comment: (Gene information with a primary key of the approved HGNC symbol for the gene. This means any gene used in the table must be approved by HGNC.) and foreign keys: .\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'gene' has columns: symbol (VARCHAR(20)): 'Approved HGNC symbol, primary key.', chr (VARCHAR(20)): 'Chromosome symbol. In the form chr##, where ## is the number or X/Y.', genesequenceid (TEXT): 'The RefSeq ID for the sequence that represents this gene, starts with \"NG_\". No version suffix.', proteinsequenceid (TEXT): 'The RefSeq ID for the sequence that represents the protein product of this gene, starts with \"NP_\". No version suffix.', chromosequenceid (TEXT): 'The RefSeq ID for the sequence that represents the chromosome this gene is on, starts with \"NC_\". No version suffix.', mrnasequenceid (TEXT): 'The RefSeq ID for the sequence that represents the translation of this gene, starts with \"NM_\". No version suffix.', hgncid (TEXT): 'The HGNC numerical ID number for this gene prefixed by \"HGNC:\"', ncbiid (TEXT): 'The NCBI Gene (Entrez) ID number for this gene', ensemblid (TEXT): 'The Ensembl ID for this gene', pharmgkbid (TEXT): 'The ID for this gene in PharmGKB.', frequencymethods (TEXT): 'Text documentation of the methods and caveats for allele frequency data', lookupmethod (TEXT): 'The way to lookup information about diplotypes of this gene, should use ACTIVITY_SCORE, PHENOTYPE, or ALLELE_STATUS', version (INTEGER), notesondiplotype (TEXT): 'Optional notes about the diplotypes of the gene', url (TEXT), functionmethods (TEXT), notesonallelenaming (TEXT), includephenotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display phenotype-level frequencies. a true value does not imply existence frequency data', includediplotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display diplotype-level frequencies. a true value does not imply existence frequency data', with comment: (Gene information with a primary key of the approved HGNC symbol for the gene. This means any gene used in the table must be approved by HGNC.) and foreign keys: .\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'gene' has columns: symbol (VARCHAR(20)): 'Approved HGNC symbol, primary key.', chr (VARCHAR(20)): 'Chromosome symbol. In the form chr##, where ## is the number or X/Y.', genesequenceid (TEXT): 'The RefSeq ID for the sequence that represents this gene, starts with \"NG_\". No version suffix.', proteinsequenceid (TEXT): 'The RefSeq ID for the sequence that represents the protein product of this gene, starts with \"NP_\". No version suffix.', chromosequenceid (TEXT): 'The RefSeq ID for the sequence that represents the chromosome this gene is on, starts with \"NC_\". No version suffix.', mrnasequenceid (TEXT): 'The RefSeq ID for the sequence that represents the translation of this gene, starts with \"NM_\". No version suffix.', hgncid (TEXT): 'The HGNC numerical ID number for this gene prefixed by \"HGNC:\"', ncbiid (TEXT): 'The NCBI Gene (Entrez) ID number for this gene', ensemblid (TEXT): 'The Ensembl ID for this gene', pharmgkbid (TEXT): 'The ID for this gene in PharmGKB.', frequencymethods (TEXT): 'Text documentation of the methods and caveats for allele frequency data', lookupmethod (TEXT): 'The way to lookup information about diplotypes of this gene, should use ACTIVITY_SCORE, PHENOTYPE, or ALLELE_STATUS', version (INTEGER), notesondiplotype (TEXT): 'Optional notes about the diplotypes of the gene', url (TEXT), functionmethods (TEXT), notesonallelenaming (TEXT), includephenotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display phenotype-level frequencies. a true value does not imply existence frequency data', includediplotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display diplotype-level frequencies. a true value does not imply existence frequency data', with comment: (Gene information with a primary key of the approved HGNC symbol for the gene. This means any gene used in the table must be approved by HGNC.) and foreign keys: .\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'gene' has columns: symbol (VARCHAR(20)): 'Approved HGNC symbol, primary key.', chr (VARCHAR(20)): 'Chromosome symbol. In the form chr##, where ## is the number or X/Y.', genesequenceid (TEXT): 'The RefSeq ID for the sequence that represents this gene, starts with \"NG_\". No version suffix.', proteinsequenceid (TEXT): 'The RefSeq ID for the sequence that represents the protein product of this gene, starts with \"NP_\". No version suffix.', chromosequenceid (TEXT): 'The RefSeq ID for the sequence that represents the chromosome this gene is on, starts with \"NC_\". No version suffix.', mrnasequenceid (TEXT): 'The RefSeq ID for the sequence that represents the translation of this gene, starts with \"NM_\". No version suffix.', hgncid (TEXT): 'The HGNC numerical ID number for this gene prefixed by \"HGNC:\"', ncbiid (TEXT): 'The NCBI Gene (Entrez) ID number for this gene', ensemblid (TEXT): 'The Ensembl ID for this gene', pharmgkbid (TEXT): 'The ID for this gene in PharmGKB.', frequencymethods (TEXT): 'Text documentation of the methods and caveats for allele frequency data', lookupmethod (TEXT): 'The way to lookup information about diplotypes of this gene, should use ACTIVITY_SCORE, PHENOTYPE, or ALLELE_STATUS', version (INTEGER), notesondiplotype (TEXT): 'Optional notes about the diplotypes of the gene', url (TEXT), functionmethods (TEXT), notesonallelenaming (TEXT), includephenotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display phenotype-level frequencies. a true value does not imply existence frequency data', includediplotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display diplotype-level frequencies. a true value does not imply existence frequency data', with comment: (Gene information with a primary key of the approved HGNC symbol for the gene. This means any gene used in the table must be approved by HGNC.) and foreign keys: .\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'gene' has columns: symbol (VARCHAR(20)): 'Approved HGNC symbol, primary key.', chr (VARCHAR(20)): 'Chromosome symbol. In the form chr##, where ## is the number or X/Y.', genesequenceid (TEXT): 'The RefSeq ID for the sequence that represents this gene, starts with \"NG_\". No version suffix.', proteinsequenceid (TEXT): 'The RefSeq ID for the sequence that represents the protein product of this gene, starts with \"NP_\". No version suffix.', chromosequenceid (TEXT): 'The RefSeq ID for the sequence that represents the chromosome this gene is on, starts with \"NC_\". No version suffix.', mrnasequenceid (TEXT): 'The RefSeq ID for the sequence that represents the translation of this gene, starts with \"NM_\". No version suffix.', hgncid (TEXT): 'The HGNC numerical ID number for this gene prefixed by \"HGNC:\"', ncbiid (TEXT): 'The NCBI Gene (Entrez) ID number for this gene', ensemblid (TEXT): 'The Ensembl ID for this gene', pharmgkbid (TEXT): 'The ID for this gene in PharmGKB.', frequencymethods (TEXT): 'Text documentation of the methods and caveats for allele frequency data', lookupmethod (TEXT): 'The way to lookup information about diplotypes of this gene, should use ACTIVITY_SCORE, PHENOTYPE, or ALLELE_STATUS', version (INTEGER), notesondiplotype (TEXT): 'Optional notes about the diplotypes of the gene', url (TEXT), functionmethods (TEXT), notesonallelenaming (TEXT), includephenotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display phenotype-level frequencies. a true value does not imply existence frequency data', includediplotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display diplotype-level frequencies. a true value does not imply existence frequency data', with comment: (Gene information with a primary key of the approved HGNC symbol for the gene. This means any gene used in the table must be approved by HGNC.) and foreign keys: .\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "> Table desc str: Table 'allele' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER): 'The version number, iterates on modification', genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', functionalstatus (TEXT): 'The functional phenotype of this allele', clinicalfunctionalstatus (TEXT): 'The functional phenotype of this allele used for clinical systems', clinicalfunctionalsubstrate (TEXT): 'Allele clinical function substrate specificity, optional', activityvalue (VARCHAR(50)): 'Descriptor of activity score, optional', definitionid (INTEGER): 'The reference to the definition for this allele', citations (ARRAY): 'An array of PubMed IDs use as citations for this functional assignment', strength (TEXT): 'The strength of evidence', functioncomments (TEXT): 'Comments this functional assignment of this allele', findings (TEXT), frequency (JSONB), inferredfrequency (BOOLEAN): 'Indicates (with \"true\") that the frequencies for this allele have been inferred by an algorithm that aggregates the frequencies for other alleles of this gene and is not based on direct data, defaults to false, should only be 1 per gene', with comment: (An allele of a gene) and foreign keys: ['definitionid'] -> allele_definition.['id'], ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'allele_definition' has columns: id (INTEGER): 'A synthetic numerical ID, auto-assigned, primary key', version (INTEGER), genesymbol (VARCHAR(50)): 'The HGNC symbol of the gene the allele is for, required', name (TEXT): 'The name of this allele, required', pharmvarid (VARCHAR(50)): 'The PharmVar core allele ID for this allele', matchesreferencesequence (BOOLEAN): 'Indicates (with \"true\") the variants that define the named allele will match the reference sequence for the gene, defaults to false, should only be 1 per gene', structuralvariation (BOOLEAN): 'Indicates whether this allele definition is a structural variant, default is false. If true, look in the pharmvarid field for the PharmVar ID that defines the structural variation.', with comment: (The definition for an allele of a gene) and foreign keys: ['genesymbol'] -> gene.['symbol'].\n",
      "\n",
      "Table 'pair' has columns: pairid (INTEGER): 'A synthetic numerical id, automatically assigned, primary key', genesymbol (VARCHAR(20)): 'The HGNC symbol of the gene in this pair, required', drugid (VARCHAR(20)): 'The ID of the drug in this pair, required', guidelineid (INTEGER): 'The ID of a guideline this pair is described in, optional', usedforrecommendation (BOOLEAN): 'Whether the gene is used for recommendation lookup for the drug if this pair is part of a guideline, default false', version (INTEGER): 'The version number, iterates on modification', cpiclevel (VARCHAR(5)): 'The CPIC level of this pair, required', pgkbcalevel (VARCHAR(5)): 'The top level of PharmGKB Clinical Annotation for this pair, optional', pgxtesting (TEXT): 'The top level of PGx testing recommendation from PharmGKB label annotations, optional', citations (ARRAY): 'The PMID citations in an array for this pair, optional', removed (BOOLEAN): 'Has this pair been \"removed\", eg is no longer a valid pair, default false', removeddate (DATE): 'If removed, when was this pair \"removed\"', removedreason (TEXT): 'If removed, why was this pair \"removed\"', with comment: (A pair of a gene and a drug that is notable to CPIC) and foreign keys: ['drugid'] -> drug.['drugid'], ['genesymbol'] -> gene.['symbol'], ['guidelineid'] -> guideline.['id'].\n",
      "\n",
      "Table 'gene' has columns: symbol (VARCHAR(20)): 'Approved HGNC symbol, primary key.', chr (VARCHAR(20)): 'Chromosome symbol. In the form chr##, where ## is the number or X/Y.', genesequenceid (TEXT): 'The RefSeq ID for the sequence that represents this gene, starts with \"NG_\". No version suffix.', proteinsequenceid (TEXT): 'The RefSeq ID for the sequence that represents the protein product of this gene, starts with \"NP_\". No version suffix.', chromosequenceid (TEXT): 'The RefSeq ID for the sequence that represents the chromosome this gene is on, starts with \"NC_\". No version suffix.', mrnasequenceid (TEXT): 'The RefSeq ID for the sequence that represents the translation of this gene, starts with \"NM_\". No version suffix.', hgncid (TEXT): 'The HGNC numerical ID number for this gene prefixed by \"HGNC:\"', ncbiid (TEXT): 'The NCBI Gene (Entrez) ID number for this gene', ensemblid (TEXT): 'The Ensembl ID for this gene', pharmgkbid (TEXT): 'The ID for this gene in PharmGKB.', frequencymethods (TEXT): 'Text documentation of the methods and caveats for allele frequency data', lookupmethod (TEXT): 'The way to lookup information about diplotypes of this gene, should use ACTIVITY_SCORE, PHENOTYPE, or ALLELE_STATUS', version (INTEGER), notesondiplotype (TEXT): 'Optional notes about the diplotypes of the gene', url (TEXT), functionmethods (TEXT), notesonallelenaming (TEXT), includephenotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display phenotype-level frequencies. a true value does not imply existence frequency data', includediplotypefrequencies (BOOLEAN): 'if allele frequencies exist for this gene, whether to calculate and display diplotype-level frequencies. a true value does not imply existence frequency data', with comment: (Gene information with a primary key of the approved HGNC symbol for the gene. This means any gene used in the table must be approved by HGNC.) and foreign keys: .\n",
      "\n",
      "Table 'drug' has columns: drugid (VARCHAR(20)): 'A unique identifier for this drug in the form \"source:id\" where source is some outside knowledge resource and id is their identifier, primary key', name (TEXT): 'The generic name for this drug, lower-cased, required', pharmgkbid (VARCHAR(20)): 'The PharmGKB ID for this drug, optional', rxnormid (VARCHAR(20)): 'The RxNorm ID for this drug, optional', drugbankid (VARCHAR(20)): 'The DrugBank ID for this drug, optional', atcid (ARRAY): 'One or more ATC IDs for this drug in an array, optional', umlscui (VARCHAR(20)): 'The UMLS Concept Unique ID for this drug, optional', flowchart (TEXT): 'URL for the flow chart image of this drug', version (INTEGER), guidelineid (INTEGER): 'The guideline this drug is part of a guideline, null means not a part of any guideline', with comment: (A clinically-used drug.) and foreign keys: ['guidelineid'] -> guideline.['id'].\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "HTTP Request: POST http://oai.hconeai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletionResponse(text='The SNP IDs associated with CYP2C19 *17 are rs12248560 and rs3758581.', additional_kwargs={}, raw={'id': 'chatcmpl-9A2XonaOD90Gj40dpFqcxKSRrvavi', 'choices': [Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='The SNP IDs associated with CYP2C19 *17 are rs12248560 and rs3758581.', role='assistant', function_call=None, tool_calls=None))], 'created': 1712180088, 'model': 'gpt-3.5-turbo-0125', 'object': 'chat.completion', 'system_fingerprint': 'fp_b28b39ffa8', 'usage': CompletionUsage(completion_tokens=23, prompt_tokens=2295, total_tokens=2318)}, logprobs=None, delta=None)"
      ]
     },
     "execution_count": 426,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = joint_query(\"What are the SNP IDs associated with CYP2C19 *17?\")\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The SNP IDs associated with CYP2C19 *17 are rs12248560 and rs3758581.'"
      ]
     },
     "execution_count": 427,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.text"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pgx_llm_rag",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
